### PR TITLE
[MDS-4597] major mine application all projects view

### PR DESCRIPTION
**Pull request by areyeslo**
## Objective 

[MDS-4597](https://bcmines.atlassian.net/browse/MDS-4597)

Create a Search page to search for projects by stage. If a project has different stages (Project Summary, IRT, Major Mine Application), then the most updated stage is shown.

image

Status code changes:
- Project Summary:
  -Deleted status_codes: OPN- Open, CLD- Closed, IAT-Inactive, UNP- Under review- with proponents
  -Updated description for UNR- Under review to reviewers to Under review

- IRT:
 -Updated PRG- In progress to DFT- Draft, REC- Pending Review to SUB- Submitted
 -Updated description for APV- Review Complete to APV- Approved
 
- Major Mine Application:
 -Updated REC- Pending Review to SUB- Sbumitted
 -Updated UNR- In review with reviewers to Under review

Sorters:
Currently, there are some fields that cannot be sorted on the table:
- Stage
- Review Status
- Updated Date
- Commodity

Each project can have multiple stages (Project Summary, IRT, MMA) and when Front End sends the request to sort by specific sort_field, sqlalchemy includes it in the query. However, it is until getting the results that we can obtain the most updated stage and pick this one from the three records (Project Summary, IRT, MMA) returned by the query. Then, sorting by stage, review_status or updated_date is not possible at the moment with the current logic using sqlalchemy.
Commodity field is obtained during transformation of row data.

### CODE REVIEW

File: services/core-api/app/api/projects/project/resources/project.py [hunk 1]
```diff
@@ -72,3 +81,211 @@ def get(self):
...
+        sort_models = {
+            'project_title': 'Project',
+            'project_id': 'Project',
+            'mine_commodity_code': 'MineTypeDetail',
+            'status_code':'ProjectSummary',
+            'status_code':'InformationRequirementsTable',
+            'status_code': 'MajorMineApplication',
+            'update_timestamp': 'ProjectSummary',
+            'update_timestamp': 'InformationRequirementsTable',
+            'update_timestamp': 'InformationRequirementsTable'
```
Comment (BenHarrington) [inline services/core-api/app/api/projects/project/resources/project.py]: I think this third "update_timestamp" was meant to be MajorMineApplication?

File: services/core-api/app/api/projects/response_models.py [hunk 2]
```diff
@@ -214,3 +214,32 @@ def format(self, value):
...
+    'ProjectMineList',{
+        'stage': fields.String,
+        'id': fields.Integer,
+        'guid': fields.String,
+        'project_title': fields.String,
+        'project_id': fields.String,
+        'project_guid': fields.String,
+        'mrc_review_required': fields.Boolean,
+        'status_code': fields.String,
+        'contacts': fields.List(fields.Nested(PROJECT_SUMMARY_CONTACT_MODEL)),
```
Comment (BenHarrington) [inline services/core-api/app/api/projects/response_models.py]: There is a PROJECT_CONTACT_MODEL I believe that should be used here. I think the PS Contact has been retired, but would need to double check that.

File: services/core-web/common/constants/actionTypes.js [hunk 3]
```diff
@@ -101,6 +101,7 @@ export const STORE_REQUIREMENTS = "STORE_REQUIREMENTS";
 export const CLEAR_REQUIREMENTS = "CLEAR_REQUIREMENTS";
 export const STORE_MAJOR_MINES_APPLICATION = "STORE_MAJOR_MINES_APPLICATION";
 export const CLEAR_MAJOR_MINES_APPLICATION = "CLEAR_MAJOR_MINES_APPLICATION";
+export const STORE_PROJECT_VIEW_ALL_TABLE = "STORE_PROJECT_VIEW_ALL_TABLE";
```
Comment (BenHarrington) [inline services/core-web/common/constants/actionTypes.js:106]: Any reason a "CLEAR..." variable wasn't provided as well here?

File: services/core-web/src/components/navigation/NavBar.js [hunk 4]
```diff
@@ -314,6 +314,18 @@ export class NavBar extends Component {
               </Link>
             </Col>
           </Row>
+          <Row>
+            <Col span={24}>
+              <Link to={router.MAJOR_PROJECTS_DASHBOARD.route}>
```
Comment (BenHarrington) [inline services/core-web/src/components/navigation/NavBar.js]: Are we putting this dashboard behind a feature/ENV wall first for testing or is it straight to Prod?

File: services/core-web/src/components/navigation/NavBar.js [hunk 5]
```diff
@@ -413,6 +425,11 @@ export class NavBar extends Component {
           <button type="button">Notices of Work</button>
         </Link>
       </Menu.Item>
+      <Menu.Item key="browse-major-projects" className="navbar-dropdown-menu-item">
+        <Link to={router.MAJOR_PROJECTS_DASHBOARD.route}>
```
Comment (BenHarrington) [inline services/core-web/src/components/navigation/NavBar.js]: Same as previous comment.

File: services/core-api/app/api/projects/project/resources/project.py (hunk 1 above)
Comment (areyeslo) [reply services/core-api/app/api/projects/project/resources/project.py]: That's correct. Thank you for noticing

File: services/core-api/app/api/projects/response_models.py (hunk 2 above)
Comment (areyeslo) [reply services/core-api/app/api/projects/response_models.py]: Agree, contacts has to be the same as PROJECT_MODEL.

File: services/core-web/src/components/navigation/NavBar.js (hunk 4 above)
Comment (areyeslo) [reply services/core-web/src/components/navigation/NavBar.js]: Good idea, going to put the behind feature flag

**Review by BenHarrington — COMMENTED** [review]
Pagination controls do not seem to be setup correctly: 
image

The previous/next buttons are disabled even though we have more than 1 page of results. Additionally the result count is 0.

**Review by BenHarrington — COMMENTED** [review]
I notice only a couple columns on the table offer sort: 
image

Was this a miss or intentional? It seems a bit inconsistent between the screenshot on the ticket and mockup, but the AC in the ticket does say these fields should be sortable(unless those reqs changed).

**Review by BenHarrington — COMMENTED** [review]
I notice on the "Application Statuses" filter control that we have over 10 eligible statuses. Were we not consolidating those to simplify things? Perhaps there is another static content selector that needs to be updated, unless we are still intending to have all of those statuses as eligible.

File: services/core-api/app/api/projects/namespace.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
 from flask_restplus import Namespace
 
-from app.api.projects.project.resources.project import ProjectResource, ProjectListResource
+from app.api.projects.project.resources.project import ProjectResource, ProjectListResource, MajorProjectListDashboardResource
```
Comment (JustinMacaulay) [inline services/core-api/app/api/projects/namespace.py]: should the name of the new resource be inline with the others? ProjectListDashboardResource?

**Review by JustinMacaulay — COMMENTED** [review]
Just a comment around the naming.

**Review by BenHarrington — APPROVED** [review]
Great job on this huge task Arturo. Very good work handling the complicated sort/filtering on the backend.

**Pull request merged**